Added support to password manager pro (PMP)

monwolf opened the pull request:
##### SUMMARY
Added support to password manager pro (PMP). This new module allow upgrade an inventory existing on Password Manager PRO and retrieve the password of a given user for a host.

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New Module Pull Request


##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
module/plugin/web_infrastructure/pmp.py
##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible --version
ansible 2.3.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.5 (default, Nov  6 2016, 00:28:07) [GCC 4.8.5 20150623 (Red Hat 4.8.5-11)]

```

Berni69 committed:
Added support to password manager pro (PMP)

Added support to password manager pro (PMP). This new module allow upgrade an inventory existing on Password Manager PRO and retrieve the password of a given user for a host.

Berni69 committed:
Update requirements.txt

Berni69 committed:
Escape \u in docs

Escape \u in docs

Berni69 committed:
changed to use ansible convenctions

Berni69 committed:
update metadata info

Berni69 committed:
updated docs

Berni69 committed:
Add RETURN to docs

Berni69 committed:
change metadata info..

monwolf commented:
shipit

jctanner commented on requirements.txt:
```diff
@@ -8,3 +8,4 @@ PyYAML
 paramiko
 cryptography
 setuptools
+requests
```
modules should never affect the install requirements for the controller.

monwolf replied:
You are right I put it on earlier version of module before using your open_url implementation

Berni69 committed:
Update requirements.txt

abadger commented on lib/ansible/modules/web_infrastructure/pmp.py:
```diff
@@ -0,0 +1,296 @@
+    }
+
+    use_proxy = False
+    validate_certs = False
```
Doesn't really matter since you're not likely to instantiate more than one PasswordManagerPro object in this short script but these should be instance variables, not class attributes.  Move setting of default values into the ```__init__```

abadger replied:
I'd also default to True for validate_certs as that is a safer default.

abadger commented on lib/ansible/modules/web_infrastructure/pmp.py:
```diff
@@ -0,0 +1,296 @@
+            strr = strr.replace('{' + key + '}', value)
+        return strr
+
+    def __init__(self, args=None):
```
Use proper keyword args here instead of passing in an opaque dict.  That way python can verify that the proper arguments are being passed in:
``` python
def __init__(self, token="", host="", port="", use_proxy=False, validate_certs=True)
    self.data = {'TOKEN': token, 'HOST': host, 'PORT': port}
    self.use_proxy = use_proxy
    self.validate_certs = validate_certs
```

abadger commented on lib/ansible/modules/web_infrastructure/pmp.py:
```diff
@@ -0,0 +1,296 @@
+            if 'validate_certs' in args:
+                self.validate_certs = args['validate_certs']
+
+    def getResources(self):
```
This is just style but we try to follow standard python naming conventions.  So function and method names should be underscore separated like this: ```get_resources```.

abadger commented on lib/ansible/modules/web_infrastructure/pmp.py:
```diff
@@ -0,0 +1,296 @@
+    pmp = PasswordManagerPro(auth_pmp)
+    status = pmp.createResource(params['resource_name'], params['account_name'], params['password'], params['owner'],
+                                group=params['group'])
+    return {'status': status}
```
Since the very first thing that both of the above do is create a ```PasswordManagerPro``` object and then they just lightly wrap a method call on that object, it seems like they should really be integrated into the methods they're calling.

abadger commented on lib/ansible/modules/web_infrastructure/pmp.py:
```diff
@@ -0,0 +1,296 @@
+    missing = []
+    for parm in OP_REQUIRED[op]:
+        if not module.params[parm]:
+            missing.append(parm)
```
You can use ```required_if``` which is built into AnsibleModule instead of having your own code.  Look at https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/cloudstack/cs_storage_pool.py#L489  as a random example of using that.

abadger commented on lib/ansible/modules/web_infrastructure/pmp.py:
```diff
@@ -0,0 +1,296 @@
+        thismod = sys.modules[__name__]
+        method = getattr(thismod, op)
+
+        ret = method(pmp_auth, module.params)
```
This looks needlessly generic.  why not just do:
``` python
if op == 'get_password':
    get_password([...])
elif op == 'create_host':
    create_host([..])
```

If you think you're going to have a lot of operations, you can have a lookup table:
```
OPERATIONS = {'get_password': get_password, 'create_host': create_host}
method = OPERATIONS[op]([...])
```

but for just a few operations I'm not sure I would do that... it obscures what the code is doing in any given situation to code it like that.

monwolf replied:
My plan is to add more operations on  the future, such as destroy an account, add an extra account for a resource, destroy a host...

But at this moment I have implemented this.

abadger commented:
A few higher level things I notice about this:

It has an "operations" parameter.  We'd rather our modules establish "state" rather than perform "actions".  describing a state and having the module ensure that state is idempotent.  Performing an action is not necessarily idempotent.  This could just be a matter of using a different name for the parameters.  You'll know better than I.

* get_password isn't establishing a state at all.  It's returning information.  We've discussed this in the past but I'm not sure what the verdict was.  Perhaps that that sort of thing should be returned as a fact?  @bcoca ?

abadger commented:
On thinkng about get_password.... maybe that portion should be a lookup plugin rather than a module?

abadger commented on lib/ansible/modules/web_infrastructure/pmp.py:
```diff
@@ -0,0 +1,296 @@
+version_added: "2.5"
+short_description: Create and modify accounts and groups in Password Manager Pro (PMP)
+description:
+  - Create and modify accounts and groups in Password Manager Pro (PMP).
```
Maybe a little more information about what password manager pro would be worthwhile to have here.

abadger commented:
If we change operations to state, It probably makes sense to have state= (present|absent)   state present makes sure that the entry exists.  state absent makes sure that the entry does not exist.

thaumos commented:
Having a quick glance at this....

- I'd like to ask for this to be two modules.  One for accounts, one for groups. 
- Have operation be state: absent, present
- Have a lookup plugin for getting password information
- Finally, I'd have a separate lookup plugin to get password info.

monwolf commented:
I'm not sure to understand the diference between plugin and module, could you explain me what's the difference? 

@jctanner nope :)

Berni69 committed:
Modified to follow recomendations.

simplified main, follow PEP8 inside class, changed docs, changed class atributes for instance variables and changed default behaviour of SSL and proxy.

monwolf commented:
@abadger when I understand what's the difference about module/plugin I will think about change the operations to states.

abadger commented:
http://docs.ansible.com/ansible/latest/playbooks_lookups.html

monwolf commented:
@abadger  Thanks for the link.
Before start to coding again I have an stupid question: 
There are some path in ansible project to store a  class shared between a module and a plugin? have I to replicate the PasswordManagerClass inside both items? how would you do that ?

Commit:
added pmp plugin to recover passwords, added support for absent/present orders

Committer: berni69 <berni.emerald@gmail.com>

monwolf committed:
Added state suppor for module, created plugin pmp to retrieve passwords

monwolf committed:
Modified to follow the coding standard
Committer: berni69 <berni.emerald@gmail.com>

monwolf committed:
delete extra '

monwolf commented:
@abadger  My changes are ok now, but there are a test failing due to a timeout downloading an rpm.

ricalenil commented:
Hello, I'm testing this module, and I'm having some errors.

When adding a resource, with an owner different to the api user, ansible is getting a timeout in the response, but the resource or user is being added to PMP successfuly.

This is my playbook..

```
- name: Hello World Sample
  hosts: localhost
  tasks:
    - name: Create Passwords
      pmp:
        uri: 'localhost'
        port: '7272'
        state: 'present'
        token: '864F4D49-1D11-4362-AFC0-B13941F061DB'
        resource_name: 'test'
        # group: 'TEST_GROUP'
        account_name: '{{ item }}'
        password: 'test'
        owner: 'admin'
        validate_certs: False
        use_proxy: False
      register: result
      with_items:
        - user1
        - user2
        - user3
```
This is the output.
```
TASK [Create Passwords] *******************************************************************************************************************************************
ok: [localhost] => (item=user1)
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: ssl.SSLError: ('The read operation timed out',)
failed: [localhost] (item=user2) => {"changed": false, "item": "user2", "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/rfletcher/ansible_Phhid5/ansible_module_pmp.py\", line 262, in <module>\n    main()\n  File \"/tmp/rfletcher/ansible_Phhid5/ansible_module_pmp.py\", line 255, in main\n    module.params['owner'], group=module.params['group'])\n  File \"/tmp/rfletcher/ansible_Phhid5/ansible_module_pmp.py\", line 214, in create_resource\n    use_proxy=self.use_proxy)\n  File \"/tmp/rfletcher/ansible_Phhid5/ansible_modlib.zip/ansible/module_utils/urls.py\", line 957, in open_url\n  File \"/usr/lib/python2.7/urllib2.py\", line154, in urlopen\n    return opener.open(url, data, timeout)\n  File \"/usr/lib/python2.7/urllib2.py\", line 429, in open\n    response = self._open(req, data)\n  File \"/usr/lib/python2.7/urllib2.py\", line 447, in _open\n    '_open', req)\n  File \"/usr/lib/python2.7/urllib2.py\", line 407, in _call_chain\n    result = func(*args)\n  File \"/tmp/rfletcher/ansible_Phhid5/ansible_modlib.zip/ansible/module_utils/urls.py\", line 417, in https_open\n  File \"/usr/lib/python2.7/urllib2.py\", line 1201, in do_open\n    r = h.getresponse(buffering=True)\n  File \"/usr/lib/python2.7/httplib.py\", line 1136, in getresponse\n    response.begin()\n  File\"/usr/lib/python2.7/httplib.py\", line 453, in begin\n    version, status, reason = self._read_status()\n  File \"/usr/lib/python2.7/httplib.py\", line 409, in _read_status\n    line = self.fp.readline(_MAXLINE + 1)\n  File \"/usr/lib/python2.7/socket.py\", line 480, in readline\n    data = self._sock.recv(self._rbufsize)\n  File \"/usr/lib/python2.7/ssl.py\", line 756, in recv\n    return self.read(buflen)\n  File \"/usr/lib/python2.7/ssl.py\", line 643, in read\n    v = self._sslobj.read(len)\nssl.SSLError: ('The read operation timed out',)\n", "module_stdout": "", "msg": "MODULE FAILURE", "rc": 0}
```

Regards,

berni69 commented:
It seems that it is a problem with your PMP server, I have been using this module for 2 months assigning a different owner to the API user and I have never before seen a timeout in a PMP response.

Maybe you could run ansible with the flags -vvv and ANSIBLE_KEEP_REMOTE_FILES=1, in order to get extra debug info. 

With ANSIBLE_KEEP_REMOTE_FILES=1 ansible won't remove the temporally playbook generated inside the remote host, after that you could try to print the request that the plugin is trying to do and launch it manually to see if the request is wrong or not.

Furthermore, with this request you will be able to open a ticket with the support of pmp in order to get extra information from they.

mavit commented on lib/ansible/modules/web_infrastructure/pmp.py:
```diff
@@ -0,0 +1,262 @@
+  uri:
+    required: true
+    description:
+      - Base URI for the PMP instance.
```
Reading the code, this appears to expect a hostname and not a URI.

mavit commented on lib/ansible/modules/web_infrastructure/pmp.py:
```diff
@@ -0,0 +1,262 @@
+    required: false
+    default:  7272
+    description:
+      - Base URI for the PMP instance.
```
This is a port number, not a URI.

mavit commented on lib/ansible/modules/web_infrastructure/pmp.py:
```diff
@@ -0,0 +1,262 @@
+    choices: [ present, absent ]
+    description:
+      -  Create or delete resource.
+
```
Inconsistent number of blank lines between blocks.

mavit commented on lib/ansible/modules/web_infrastructure/pmp.py:
```diff
@@ -0,0 +1,262 @@
+    required: false
+    description:
+      - Set if open_url must use proxy
+  validate_certs:
```
Should document that the default is `True`.

mavit commented on lib/ansible/modules/web_infrastructure/pmp.py:
```diff
@@ -0,0 +1,262 @@
+        argument_spec=dict(
+            uri=dict(required=True),
+            port=dict(required=False, default="7272"),
+            state=dict(choices=['present', 'absent'], required=True),
```
In most modules, `state` is optional and defaults to `present`.  It would be easier for users if this module stuck to that convention.

mavit commented on lib/ansible/modules/web_infrastructure/pmp.py:
```diff
@@ -0,0 +1,262 @@
+class PasswordManagerPro:
+    def format(self, strr):
+        for key, value in self.data.items():
+            strr = strr.replace('{' + key + '}', value)
```
It's safer to quote unvalidated strings with, e.g., [`urllib.quote`](https://docs.python.org/2.7/library/urllib.html?#urllib.quote) before concatenating them into a URL.

mavit commented on lib/ansible/plugins/lookup/pmp.py:
```diff
@@ -0,0 +1,80 @@
+    short_description: Recover password account from Password Manager Pro (PMP)
+    description:
+      - Use this module to  requests passwords from an specific account to PMP
+    options:
```
Looks like this was copy-and-pasted from the LastPass lookup plugin, and hasn't yet been updated for Password Manager Pro.

mavit commented on lib/ansible/modules/web_infrastructure/pmp.py:
```diff
@@ -0,0 +1,262 @@
+short_description: Create and delete resources in Password Manager Pro (PMP)
+description:
+  - This allows to create and delete a resource with an account inside Password Manager Pro (PMP).
+
```
It would be helpful to add a [`notes`](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#ansible-metadata-block) section that mentions to the user that they should use the lookup plugin for retrieving existing password records from Password Manager Pro.

yaplej commented:
How is this going?  I am very interested in seeing this merged.

mavit commented:
@yaplej We’ve not heard from @monwolf for a while, so it might need someone else to pick this up and finish it off.  I’m afraid I’m no-longer working for the client who used Password Manager Pro, so that won’t be me.

@ansibot needs_contributor

yaplej commented:
@mavit I might pick this up.  I'm currently working on a Cherwell API dynamic inventory script.  Hope to have that working in a week or two.

mgaruccio commented:
Does anyone know what's still needed to get this merged?  Or would it need to be abstracted into a collection now?  Can probably put in the work if it doesn't need a complete rewrite.

sivel commented:
We are no longer accepting new modules and plugins in this repository. This feature would need to be re-submitted to https://github.com/ansible-collections/community.general to be considered for merge.

Akasurde commented:
Closing as per above.
